feat: add `stats/base/ndarray/dcovarmtk` (#7692)
stdlib-js/stdlib: lib/node_modules/@stdlib/stats/base/ndarray/dcovarmtk/README.md

<!--

@license Apache-2.0

Copyright (c) 2025 The Stdlib Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

-->

# dcovarmtk

> > Calculate the [covariance][covariance] of two one-dimensional double-precision floating-point ndarrays provided known means and using a one-pass textbook algorithm.

Author: Is this definition correct?

Reviewer: Yes, this looks fine to me.

```

<!-- </equation> -->

The use of the term `n-1` is commonly referred to as Bessel's correction. Depending on the characteristics of the population distributions, other correction factors (e.g., `n-1.5`, `n+1`, etc) can yield better estimators.

</section>

<!-- /.intro -->

<section class="usage">

## Usage

```javascript
var dcovarmtk = require( '@stdlib/stats/base/ndarray/dcovarmtk' );
```

#### dcovarmtk( arrays )

Computes the covariance of two one-dimensional double-precision floating-point ndarrays provided known means and using a one-pass textbook algorithm.

```javascript
var Float64Array = require( '@stdlib/array/float64' );
var scalar2ndarray = require( '@stdlib/ndarray/from-scalar' );
var ndarray = require( '@stdlib/ndarray/base/ctor' );

var opts = {
    'dtype': 'float64'
};

var xbuf = new Float64Array( [ 1.0, -2.0, 2.0 ] );
var x = new ndarray( opts.dtype, xbuf, [ 3 ], [ 1 ], 0, 'row-major' );

var ybuf = new Float64Array( [ 2.0, -2.0, 1.0 ] );
var y = new ndarray( opts.dtype, ybuf, [ 3 ], [ 1 ], 0, 'row-major' );

var meanx = scalar2ndarray( 1.0/3.0, opts );
var meany = scalar2ndarray( 1.0/3.0, opts );
var correction = scalar2ndarray( 1.0, opts );

> var v = dcovarmtk( [ x, y, meanx, meany, correction ] );

Reviewer: @gururaj1512 Would you mind reordering the ndarrays here to

```
[ x, y, correction, meanx, meany ]
```

?

This ensures that the relative order of the ndarrays matches the order of the arguments in the corresponding strided array function

```
dcovarmtk( N, correction, meanx, x, strideX, meany, y, strideY )
```

Notice: `correction`, `meanx`, and `meany` order.

That applies throughout this PR.

// returns ~3.8333
```

The function has the following parameters:

-   **arrays**: array-like object containing the following ndarrays in order:

    1.  first one-dimensional input ndarray.
    2.  second one-dimensional input ndarray.

>     3.  a zero-dimensional ndarray specifying degrees of freedom adjustment. Setting this parameter to a value other than `0` has the effect of adjusting the divisor during the calculation of the [covariance][covariance] according to `N-c` where `c` corresponds to the provided degrees of freedom adjustment. When computing the population [covariance][covariance], setting this parameter to `0` is the standard choice (i.e., the provided arrays contain data constituting entire populations). When computing the unbiased sample [covariance][covariance], setting this parameter to `1` is the standard choice (i.e., the provided arrays contain data sampled from larger populations; this is commonly referred to as Bessel's correction).

Reviewer: ```suggestion
    3.  a zero-dimensional ndarray specifying the degrees of freedom adjustment. Setting this parameter to a value other than `0` has the effect of adjusting the divisor during the calculation of the [covariance][covariance] according to `N-c` where `c` corresponds to the provided degrees of freedom adjustment and `N` corresponds to the number of elements in each input ndarray. When computing the population [covariance][covariance], setting this parameter to `0` is the standard choice (i.e., the provided arrays contain data constituting entire populations). When computing the unbiased sample [covariance][covariance], setting this parameter to `1` is the standard choice (i.e., the provided arrays contain data sampled from larger populations; this is commonly referred to as Bessel's correction).
```

    4.  a zero-dimensional ndarray specifying mean of the first one-dimensional ndarray.

>     5.  a zero-dimensional ndarray specifying mean of the second one-dimensional ndarray.

Reviewer: ```suggestion
    4.  a zero-dimensional ndarray specifying the mean of the first one-dimensional ndarray.
    5.  a zero-dimensional ndarray specifying the mean of the second one-dimensional ndarray.
```

</section>

<!-- /.usage -->

<section class="notes">

## Notes

> 
> -   If provided an empty one-dimensional ndarray, the function returns `NaN`.

Reviewer: ```suggestion

-   Both input ndarrays should have the same number of elements.
-   If provided empty one-dimensional ndarrays, the function returns `NaN`.
```

</section>

<!-- /.notes -->

<section class="examples">

## Examples

<!-- eslint no-undef: "error" -->

```javascript
var discreteUniform = require( '@stdlib/random/array/discrete-uniform' );
var ndarray = require( '@stdlib/ndarray/base/ctor' );
var ndarray2array = require( '@stdlib/ndarray/to-array' );
var scalar2ndarray = require( '@stdlib/ndarray/from-scalar' );
var dcovarmtk = require( '@stdlib/stats/base/ndarray/dcovarmtk' );

// Define array options:
var opts = {
    'dtype': 'float64'
};

> // Create first one-dimensional ndarray containing pseudorandom integers drawn from a discrete uniform distribution:
> var xbuf = discreteUniform( 10, -50, 50, opts );
> var x = new ndarray( opts.dtype, xbuf, [ xbuf.length ], [ 1 ], 0, 'row-major' );
> console.log( ndarray2array( x ) );
> 
> // Create second one-dimensional ndarray containing pseudorandom integers drawn from a discrete uniform distribution:

Reviewer: ```suggestion
// Create one-dimensional ndarrays containing pseudorandom numbers:
var xbuf = discreteUniform( 10, -50, 50, opts );
var x = new ndarray( opts.dtype, xbuf, [ xbuf.length ], [ 1 ], 0, 'row-major' );
console.log( ndarray2array( x ) );

```
